firedrake/interpolation.py

```diff
@@ -278,9 +278,12 @@ def __init__(
         allow_missing_dofs=False,
         matfree=True
     ):
-        if isinstance(expr, ufl.Interpolate):
-            expr, = expr.ufl_operands
-        self.expr = expr
+        if not isinstance(expr, ufl.Interpolate):
```

Comment:
why would the expression to interpolate ever be a `ufl.Interpolate`

Reply:
Currently Interpolator is not aware of the underlying `ufl.Interpolate` expression. If we want to generate code for the matfree dual, Interpolator needs to hold both the `operand` and the `dual_arg`.  

I figured that the path of least resistance was to alway attach `ufl.Interpolate` to the `Interpolator`. This PR treats Interpolator as if it were {Zero|One|Two}FormAssembler with an assemble that turns the symbolic ufl.Interpolator into a `scalar`, `(Co)Function` or `Matrix`.

Reply:
I am mantaining backwards compatibility and allowing for an `Interpolator` to be constructed from the `operand` (in which case we internally construct the `ufl.Interpolate`), or from the `ufl.Interpolate`